src/util.cpp

```diff
@@ -435,11 +435,116 @@ static void InterpretNegativeSetting(std::string& strKey, std::string& strValue)
     }
 }
 
+/** Internal helper functions for ArgsManager */
+class ArgsManagerHelper {
+public:
+    inline static bool UseDefaultSection(const ArgsManager& am, const std::string& strArg)
+    {
+        if (am.m_strSection == CBaseChainParams::MAIN)
+            return true;
+        else if (am.m_setSectionOnlyArgs.count(strArg) == 0)
+            return true;
+        else
+            return false;
+    }
+
+    typedef std::map<std::string, std::vector<std::string>> map_args;
+
+    /** Convert regular argument into the section-specific setting */
+    inline static std::string SectionArg(const ArgsManager& am, const std::string& strArg)
+    {
+        assert(strArg.length() > 1 && strArg[0] == '-');
+        return "-" + am.m_strSection + "." + strArg.substr(1);
+    }
+
+    /** Find arguments in a map and add them to a vector */
+    inline static void AddArgs(std::vector<std::string>& vRes, const map_args& mapArgs, const std::string& strArg)
+    {
+        auto it = mapArgs.find(strArg);
+        if (it != mapArgs.end()) {
+            vRes.insert(vRes.end(), it->second.begin(), it->second.end());
+        }
+    }
+
+    /** Return true/false if an argument is set in a map, and also
+     *  return the first (or last) of the possibly multiple values it has
+     */
+    static inline bool GetArgHelper(std::string* result, const map_args& mapArgs, const std::string& strArg, bool getLast = false)
+    {
+        auto it = mapArgs.find(strArg);
+
+        if (it == mapArgs.end() || it->second.empty())
+            return false;
+
+        if (result != nullptr) {
+            if (getLast) {
+                *result = it->second.back();
+            } else {
+                *result = it->second.front();
+            }
+        }
+
+        return true;
+    }
+
+    /* Get the string value of an argument, returning true if found, or
+     * false if not found.
+     */
+    static inline bool GetArg(const ArgsManager &am, std::string* result, const std::string& strArg)
+    {
+        LOCK(am.cs_args);
+
+        if (GetArgHelper(result, am.m_mapOverrideArgs, strArg, true))
+            return true;
+
+        if (!am.m_strSection.empty()) {
+            if (GetArgHelper(result, am.m_mapConfigArgs, SectionArg(am, strArg)))
+                return true;
+        }
+
+        if (UseDefaultSection(am, strArg)) {
```

Comment:
Is it possible to log when a default section config is not used? It could be confusing for users who specify `-addnode`, `-connect`, etc in the default section, expecting the config to be picked up.

Reply:
As in "Warning: you're running on testnet, so your -addnode setting didn't get applied" ?

Reply:
Or "Warning: Setting X specified outside configuration section is only applied on mainnet"